hera_cal/frf.py

```diff
@@ -97,6 +106,312 @@ def sky_frates(uvd, keys=None, frate_standoff=0.0, frate_width_multiplier=1.0, m
+        if True, set profiles for all pols equal to the sum of the profiles of individual pols.
+    dfr: float, optional.
+        spacing of fringe-rate grid used to perform binning and percentile calc
+        in units of mHz.
+        default is None -> set to 1 / (dtime * Ntime) of uvd.
+    nfr: float, optional.
+        number of points on fringe-rate grid to perform binning and percentile calc.
+        default is None -> set to uvd.Ntimes.
+    taper: str, optional
+        taper expected for power spectrum calculations. Fringe-rates from different frequencies
+    fr_freq_skip: int, optional
+        bin fringe rates from every freq_skip channels.
+        default is 1 -> takes a long time. We recommend setting this to be larger.
+    verbose: bool, optional
+        lots of text output.
+    Returns
+    -------
+    fr_grid: np.ndarray
+        length nfr grid of fringe-rates spaced by dfr centered at zero.
+
+    profiles: Dictionary object. Maps antpairpol tuples to numpy.ndarray object
+              with the sum of the beam squared in all directions falling into
+              a particular fringe-rate bin.
+
+    """
+
+    uvb = copy.deepcopy(uvb)
+    # convert to power and healpix if necesssary
+    if uvb.beam_type == 'efield':
+        uvb.efield_to_power()
+    try:
+        uvb.to_healpix()
+    except ValueError as err:
+        warnings.warn("UVBeam object already in healpix format...")
+
+    if keys is None:
+        keys = uvd.get_antpairpols()
+
+    antpos_trf = uvd.antenna_positions  # earth centered antenna positions
+    antnums = uvd.antenna_numbers  # antenna numbers.
+
+    lat, lon, alt = uvd.telescope_location_lat_lon_alt_degrees
+    location = EarthLocation(lon=lon * units.deg, lat=lat * units.deg, height=alt * units.m)
+
+    # get topocentricl AzEl Beam coordinates.
+    npix = uvb.data_array.shape[-1]
+    nside = hp.npix2nside(npix)
+    polar, az = hp.pix2ang(nside=nside, ipix=range(npix))
+    # cancel out super-horizon beam
+    alt = np.pi / 2. - polar
+
+    # Covert AltAz coordinates of UVBeam pixels to barycentric coordinates.
+    obstime = Time(np.median(np.unique(uvd.time_array)), format='jd')
+    altaz = AltAz(obstime=obstime, location=location)
+    itrs = ITRS()
+    # coordinates of beam pixels in topocentric frame.
+    altaz_coords = SkyCoord(alt=alt * units.rad, az=az * units.rad, frame=altaz)
+    # transform beam pixels from topocentric to ITRS
+    eq_coords = altaz_coords.transform_to(itrs)
+    # get cartesian xyz unit vectors in the direction of each beam pixle.
+    eq_xyz = np.vstack([eq_coords.x, eq_coords.y, eq_coords.z])
+
+    # generate fringe_rate grid in mHz.
+    dt = SDAY_KSEC * np.median(np.diff(np.unique(uvd.time_array)))  # times in kSec
+    if nfr is None:
+        nfr = uvd.Ntimes
+    if dfr is None:
+        # if no dfr provided, set to 1 / (ntimes * dt)
+        dfr = 1. / (dt * nfr)
+
+    # build grid.
+    fr_grid = np.arange(-nfr // 2, nfr // 2) * dfr
```

Comment:
I'm not sure if this is guaranteed to be length nfr (when things are perfectly divisible, for example), but it would be if you use linspace.